distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerProcessBuilder.java

```diff
@@ -0,0 +1,188 @@
+     * Specifies the temporary directory to be used by the server process
+     */
+    public ServerProcessBuilder withTempDir(Path tempDir) {
+        this.tempDir = tempDir;
+        return this;
+    }
+
+    /**
+     * Specifies the process info to pass through to the new Java subprocess
+     */
+    public ServerProcessBuilder withProcessInfo(ProcessInfo processInfo) {
+        this.processInfo = processInfo;
+        return this;
+    }
+
+    /**
+     * Specifies the command line arguments to run Elasticsearch
+     */
+    public ServerProcessBuilder withServerArgs(ServerArgs serverArgs) {
+        this.serverArgs = serverArgs;
+        return this;
+    }
+
+    /**
+     * Specifies the JVM options to be passed to the Elasticsearch Java process
+     */
+    public ServerProcessBuilder withJvmOptions(List<String> jvmOptions) {
+        this.jvmOptions = jvmOptions;
+        return this;
+    }
+
+    /**
+     * Specifies the {@link Terminal} to use for reading input and writing output from/to the cli console
+     */
+    public ServerProcessBuilder withTerminal(Terminal terminal) {
+        this.terminal = terminal;
+        return this;
+    }
+
+    private Map<String, String> getEnvironment() {
+        Map<String, String> envVars = new HashMap<>(processInfo.envVars());
+
+        envVars.remove("ES_TMPDIR");
+        if (envVars.containsKey("LIBFFI_TMPDIR") == false) {
+            envVars.put("LIBFFI_TMPDIR", tempDir.toString());
+        }
+        envVars.remove("ES_JAVA_OPTS");
+
+        return envVars;
+    }
+
+    private List<String> getJvmArgs() {
+        Path esHome = processInfo.workingDir();
+        return List.of(
+            "--module-path",
+            esHome.resolve("lib").toString(),
+            // Special circumstances require some modules (not depended on by the main server module) to be explicitly added:
+            "--add-modules=jdk.net", // needed to reflectively set extended socket options
+            // we control the module path, which may have additional modules not required by server
+            "--add-modules=ALL-MODULE-PATH",
+            "-m",
+            "org.elasticsearch.server/org.elasticsearch.bootstrap.Elasticsearch"
+        );
+    }
+
+    private String getCommand() {
+        Path javaHome = PathUtils.get(processInfo.sysprops().get("java.home"));
+
+        boolean isWindows = processInfo.sysprops().get("os.name").startsWith("Windows");
+        return javaHome.resolve("bin").resolve("java" + (isWindows ? ".exe" : "")).toString();
+    }
+
+    /**
+     * Start a server in a new process.
+     *
+     * @return A running server process that is ready for requests
+     * @throws UserException        If the process failed during bootstrap
+     */
+    public ServerProcess start() throws UserException {
+        return start(ProcessBuilder::start);
+    }
+
+    // package private for testing
+    ServerProcess start(ProcessStarter processStarter) throws UserException {
```

Comment:
I know this is mostly a matter of style. The builder api allows to build a server process on a partially configured builder which would cause NPEs in different places. Builders, unfortunately, often obfuscate what's required and leave callers blind / having to guess, making it important to fail fast / validate the builder configuration.

This is more of a question, absolutely no objection to the code here (particularly as this is fairly isolated). I'm kinda used to null-safe coding, also by means of static code analysis. What would be the best practice / common approach on elasticsearch? It looks like that's mostly done using assertions?

Reply:
I'm glad you raised it, as I'm not particularly fond of this style of builders either. I'd rather prefer you are _not_ able to `build` (or `start`, in this case) unless you have provided all the inputs required.
That said, like you pointed out, here it is fairly isolated so I'm OK with it - but not very fond of it, and I'd also like to discuss a common approach we could follow. Let's talk about it more on slack.